composer/datasets/synthetic.py

```diff
@@ -21,6 +36,201 @@ class SyntheticDataLabelType(StringEnum):
+
+    unk_token = None
+    pad_token = "<pad>"
+
+    return SyntheticTokenizerParams(
+        tokenizer_model=tokenizers_models.BPE(unk_token=unk_token),
+        normalizer=normalizers.Lowercase(),
+        pre_tokenizer=pre_tokenizers.ByteLevel(),
+        decoder=decoders.ByteLevel(),
+        initial_alphabet=pre_tokenizers.ByteLevel.alphabet(),
+        special_tokens=[pad_token, "<endoftext>"],
+        pad_token=pad_token,
+        trainer_cls=tokenizers_trainer.BpeTrainer,
+        tokenizer_cls=GPT2Tokenizer,
+    )
+
+
+def generate_synthetic_tokenizer(tokenizer_family: str, dataset=None, vocab_size=256) -> PreTrainedTokenizer:
+    try:
+        from tokenizers import Tokenizer
+        from transformers import PreTrainedTokenizer
+    except ImportError as e:
+        raise ImportError(
+            textwrap.dedent("""\
+            Composer was installed without NLP support. To use NLP with Composer, run `pip install mosaicml[nlp]`
+            if using pip or `conda install -c conda-forge transformers tokenizers` if using Anaconda.""")) from e
+
+    # generate a synthetic dataset with reasonable defaults is none is provided
+    if dataset is None:
+        num_samples = 100
+        chars_per_sample = 128
+        column_names = ['sentence']
+        dataset = SyntheticHFDataset(num_samples=num_samples,
+                                     chars_per_sample=chars_per_sample,
+                                     column_names=column_names).generate_dataset()
+
+    # flatten dataset to get a simple stream of text for the tokenizer
+    dataset = [dataset[key] for key in dataset.column_names if key != 'idx']
+    dataset = [i for sublist in dataset for i in sublist]
+
+    if "bert" in tokenizer_family:
+        tokenizer_params = generate_bert_tokenizer_params(dataset)
+    elif "gpt2" in tokenizer_family:
+        tokenizer_params = generate_gpt2_tokenizer_params()
+    else:
+        raise ValueError(f"Synthetic tokenizers for tokenizer family {tokenizer_family} are currently unsupported.")
+
+    tokenizer = Tokenizer(tokenizer_params.tokenizer_model)
+    tokenizer.enable_padding(direction="right",
+                             pad_id=0,
+                             pad_type_id=0,
+                             pad_token=tokenizer_params.pad_token,
+                             pad_to_multiple_of=8)
+    tokenizer.normalizer = tokenizer_params.normalizer  # type: ignore
+    tokenizer.pre_tokenizer = tokenizer_params.pre_tokenizer  # type: ignore
+    tokenizer.decoder = tokenizer_params.decoder  # type: ignore
+    tokenizer_trainer = tokenizer_params.trainer_cls(
+        vocab_size=vocab_size,
+        initial_alphabet=tokenizer_params.initial_alphabet,
+        special_tokens=tokenizer_params.special_tokens,
+    )
+    tokenizer.train_from_iterator(dataset, trainer=tokenizer_trainer)
+
+    # save the tokenizer config
+    tmp_tokenizer_dir = mkdtemp()
```

Comment:
use `with tempfile.TemporaryDirectory() as tempdir` instead of `mkdtemp`...tempfile ensures that the directory is removed at the end.

Reply:
Done, this is now possible after the refactor! Earlier on, I didn't want the temporary directory being immediately deleted.